rust-lang/rust
# Add a Cargo-based build system to replace the Makefiles

This pull request adds a Cargo-based build system, in parallel to the old Makefile-based one. The two build systems should be able to function alongside each other, but it is preferred to use the new one and the Makefiles will be deleted once the new system has been thoroughly tested and is considered mature enough.

On a high-level, this pull request:
1. Adds the necessary `Cargo.toml` files to convert the crates into Cargo packages.
2. Adds a Cargo build script if building and linking native runtime libraries is required.
3. Adds a Python script `etc/build_llvm.py` to drive the LLVM build system.
4. Adds a driver script `etc/build_rust.py` to drive the whole system. This driver script will parse the command line arguments, build LLVM, and then invoke Cargo to bootstrap the compiler, build the docs, and build and run the compiler tests and library tests.

The new build system offers a number of advantages over the old one. They include:
1. Being a lot simpler and easier to understand and hack.
2. Reuses the Cargo infrastructure and makes it easier to move the built-in crates into crates.io
3. Cross-bootstrapping is as simple as doing `cargo build --target=<target-triple>`
4. Much easier to run the `libstd` tests without bootstrapping the compiler. This can save quite some time when you are developing the standard library.
5. Reduces the build-time dependencies, by not depending on `make` and other Unix tools when they are unavailable. You can now build the compiler under a standard Windows command line window, without having MSYS2 installed (see [the list in the tracking Windows issue](https://github.com/rust-lang/rfcs/issues/1061)).

There are a number of technical differences between the two build systems:
1. Instead of bootstrapping from a stage0 snapshot, we use the latest Rust nightly (which includes both `rustc` and `cargo`). Because the stage0 compiler is already a hosted compiler (instead of a freestanding one), there is no need to go through the staged api and compile a hosted stage0 compiler. Instead we will simply use nightly to compile the stage1 compiler and then use the stage1 compiler to compile the stage2 compiler.
2. The compiler executables (`rustc` and `rustdoc`) will be statically linked against the standard library and compiler crates. This is the default behaviour of Cargo. It also makes it easier to redistribute the compiler as there won't be any dylib dependencies of the executables, although it doesn't really make the distribution smaller (they have roughly the same size on release builds as I measured).
3. For release builds, the compiler and standard crates are compiled with `opt-level=3` and `lto=true`, to make them faster.
4. When building `compiler-rt`, we will not use the builtin Makefiles and instead invoke the C toolchain directly. This is much simpler and works on platforms where gnu `make` is not universally available. This requires an update to the `compiler-rt` source by adding a small header file which is missing for Windows (it is available for Linux and darwin), because the standalone mingw doesn't include it.

Current limitations of the new system:
1. Although the code is written in a way that should be able to handle MSVC, currently it won't work as Cargo cannot pass `-C link-args=-DEF:` to rustc. This requires extending Cargo and support for MSVC will come in pull requests to follow.
2. Although in theory cross-bootstrapping is as simple as invoking `cargo build --target=<target-triple>`, and the code is written to handle cross-bootstrapping properly, in practice there will be platform differences. Therefore cross-bootstrapping needs a lot more testing than it currently has.

To test, one needs to
1. Pull https://github.com/rust-lang/cargo/pull/1802 and recompile Cargo.
2. You will also need to pull https://github.com/rust-lang/compiler-rt/pull/9 if you want to build on Windows.
3. On a Unix-based system, run
   
   ``` sh
   $ ./build.sh
   ```
   
   On Windows, open a command line window and run
   
   ```
   python src\etc\build_rust.py
   ```
   
   To see a list of supported options, pass `--help` to the build script. For instance, you might want to use `--enable-debug` to build with debug profile (the default is to build with the release profile).

See also the discussion in https://github.com/rust-lang/rust/pull/26493.

r? @alexcrichton
